types/src/crypto/asymmetric_key.rs

```diff
@@ -168,39 +148,34 @@ impl AsymmetricType for SecretKey {
     }
 
     fn ed25519_from_bytes<T: AsRef<[u8]>>(bytes: T) -> Result<Self, Error> {
-        Ok(SecretKey::Ed25519(
-            ed25519_dalek::SecretKey::from_bytes(bytes.as_ref()).map_err(|_| {
-                Error::AsymmetricKey(format!(
-                    "failed to construct Ed25519 secret key.  Expected {} bytes, got {} bytes.",
-                    Self::ED25519_LENGTH,
-                    bytes.as_ref().len()
-                ))
-            })?,
-        ))
+        Ok(SecretKey::Ed25519(ed25519_dalek::SecretKey::from_bytes(
+            bytes.as_ref(),
+        )?))
     }
 
     fn secp256k1_from_bytes<T: AsRef<[u8]>>(bytes: T) -> Result<Self, Error> {
-        Ok(SecretKey::Secp256k1(
-            <[u8; Self::SECP256K1_LENGTH]>::try_from(bytes.as_ref()).map_err(|err| {
-                Error::AsymmetricKey(format!(
-                    "failed to construct secp256k1 secret key. \
-                     Expected {} bytes, got {} bytes. \
-                     Error: {}",
-                    Self::SECP256K1_LENGTH,
-                    bytes.as_ref().len(),
-                    err
-                ))
-            })?,
-        ))
-    }
-}
-
-impl AsRef<[u8]> for SecretKey {
-    fn as_ref(&self) -> &[u8] {
-        match self {
-            SecretKey::System => &[],
-            SecretKey::Ed25519(secret_key) => secret_key.as_ref(),
-            SecretKey::Secp256k1(secret_key) => secret_key,
+        Ok(SecretKey::Secp256k1(Secp256k1SecretKey::from_bytes(
+            bytes.as_ref(),
+        )?))
+    }
+}
+
+impl From<SecretKey> for Vec<u8> {
+    fn from(secret_key: SecretKey) -> Vec<u8> {
+        match secret_key {
+            SecretKey::System => Vec::new(),
+            SecretKey::Ed25519(key) => key.as_ref().into(),
+            SecretKey::Secp256k1(key) => key.to_bytes().as_slice().into(),
+        }
+    }
+}
```

Comment:
We shouldn't have this variant, only `impl From<&SecretKey> for Vec<u8> {`

Reply:
Well... maybe we can't get away with that...

Reply:
I've changed the original `impl` to `From<&Secret>` but I have kept a thin wrapper for `From<Secret>` as it matches previous code patterns. If it must be removed, I can do so

Reply:
Yeah, let's get rid of it I think.

Reply:
Oh I can't remove it cause its a trait bound on `AsymetricType`